File: C++/SourceCode/CovMat.cpp
#include <iostream>
#include <memory>
#include <armadillo>
#include "CovMat.hpp" 

using namespace std;
using namespace arma;

void CovMat::ConstructorInitialize()
{
	this->b_eigenValues = false;
	this->b_eigenVectors = false;
	this->b_norm = false;
	this->b_determinant = false;

	this->eigenValues = shared_ptr<vec> (new vec);
	this->eigenVectors = shared_ptr<mat> (new mat);
	this->inverse = nullptr;
	this->sqrtm = nullptr;
	this->invsqrtm = nullptr;
	this->expm = nullptr;
	this->logm = nullptr;
	this->powm = nullptr;
	this->currentPower = 1;
}

void CovMat::DeleteAllocatedVar()
{

}

void CovMat::Copy(const CovMat& covMat)
{
	this->matrix = covMat.matrix;
	this->matrixOrder = covMat.matrixOrder;
	this->b_eigenValues = covMat.b_eigenValues; this->eigenValues = covMat.eigenValues;
	this->b_eigenVectors = covMat.b_eigenVectors; this->eigenVectors = covMat.eigenVectors; this->eigenVectorsTranspose = covMat.eigenVectorsTranspose;
	this->b_norm = covMat.b_norm; this->norm = covMat.norm;
	this->b_determinant = covMat.b_determinant; this->determinant = covMat.determinant;

	this->inverse = covMat.inverse;
	this->sqrtm = covMat.sqrtm;
	this->invsqrtm = covMat.invsqrtm;
	this->expm = covMat.expm;
	this->logm = covMat.logm;
	this->currentPower = covMat.currentPower; this->powm = covMat.powm;
}

CovMat::CovMat(double* array, const unsigned matrixOrder)
{
	this->matrix = shared_ptr<mat> (new mat(array, matrixOrder, matrixOrder, false, true));
	this->matrixOrder = matrixOrder;
	this->ConstructorInitialize();
}

CovMat::CovMat(const mat& matrix)
{
	this->matrix = make_shared<mat> (matrix);
	this->matrixOrder = matrix.n_cols;
	this->ConstructorInitialize();
}

CovMat::CovMat(const unsigned int matrixOrder)
{
	this->matrix = shared_ptr<mat> (new mat(matrixOrder, matrixOrder));
	this->matrix->zeros();
	this->matrixOrder = matrixOrder;
	this->ConstructorInitialize();
}

CovMat::CovMat(const CovMat& covMat)
{
	this->Copy(covMat);
}

CovMat::CovMat()
{
	this->ConstructorInitialize();
}

CovMat::~CovMat()
{
	this->DeleteAllocatedVar();
}

void CovMat::ComputeEigen(bool eigenValuesOnly)
{
	if ((this->b_eigenValues)&&(this->b_eigenVectors))
		return;
	
	if ((this->b_eigenValues)&&(eigenValuesOnly))
		return;

	if (eigenValuesOnly)
	{
		*this->eigenValues = eig_sym(*this->matrix);
		this->b_eigenValues = true;
		return;
	}

	eig_sym(*this->eigenValues, *this->eigenVectors, *this->matrix);
	this->eigenVectorsTranspose = make_shared<mat> (this->eigenVectors->t());
	this->b_eigenValues = true;
	this->b_eigenVectors = true;
}

void CovMat::SetToZero()
{
	this->matrix->zeros();
	this->DeleteAllocatedVar();
	this->ConstructorInitialize();
}

void CovMat::Randomize()
{
	mat tmp(this->matrixOrder, 2*this->matrixOrder);
	tmp.randu();
	*this->matrix = tmp * tmp.t();
	this->DeleteAllocatedVar();
	this->ConstructorInitialize();
}

double CovMat::Norm()
{
	if (this->b_norm)
		return this->norm;

	this->norm = arma::norm(*this->matrix, "fro");
	this->b_norm = true; 

	return this->norm;
}

double CovMat::Determinant()
{
	if (this->b_determinant)
		return this->determinant;

	this->determinant = det(*this->matrix);
	this->b_determinant = true;

	return this->determinant;
}

CovMat& CovMat::Transpose()
{
	return *this;
}

CovMat& CovMat::Inverse()
{
	if (this->inverse != nullptr)
		return *(this->inverse);

	this->inverse = shared_ptr<CovMat> (new CovMat(inv_sympd(*this->matrix)));

	return *(this->inverse);
}

CovMat& CovMat::Sqrtm()
{
	if (this->sqrtm != nullptr)
		return *(this->sqrtm);

	this->ComputeEigen();
	
	this->sqrtm = shared_ptr<CovMat> (new CovMat(*this->eigenVectors * diagmat(sqrt(*this->eigenValues)) * *this->eigenVectorsTranspose));

	return *(this->sqrtm);
}

CovMat& CovMat::Invsqrtm()
{
	if (this->invsqrtm != nullptr)
		return *(this->invsqrtm);

	this->ComputeEigen();

	this->invsqrtm = shared_ptr<CovMat> (new CovMat(*this->eigenVectors * diagmat(pow(*this->eigenValues, -0.5)) * *this->eigenVectorsTranspose));

	return *(this->invsqrtm);
}

CovMat& CovMat::Expm()
{
	if (this->expm != nullptr)
		return *(this->expm);

	this->ComputeEigen();

	this->expm = shared_ptr<CovMat> (new CovMat(*this->eigenVectors * diagmat(exp(*this->eigenValues)) * *this->eigenVectorsTranspose));

	return *(this->expm);
}

CovMat& CovMat::Logm()
{
	if (this->logm != nullptr)
		return *(this->logm);

	this->ComputeEigen();

	this->logm = shared_ptr<CovMat> (new CovMat(*this->eigenVectors * diagmat(log(*this->eigenValues)) * *this->eigenVectorsTranspose));

	return *(this->logm);
}

CovMat& CovMat::Powm(const double power)
{
	if (power == 1)
		return *this;

	if (power == this->currentPower)
		return *(this->powm);

	this->ComputeEigen();

	this->powm = shared_ptr<CovMat> (new CovMat(*this->eigenVectors * diagmat(pow(*this->eigenValues, power)) * *this->eigenVectorsTranspose));
	this->currentPower = power;

	return *(this->powm);
}

ostream& operator << (ostream &output, const CovMat& covMat)
{ 
    output << *covMat.matrix;
    return output;            
}

double CovMat::operator () (const int nCol, const int nRow)
{
	return (*this->matrix)(nCol, nRow);
}

CovMat& CovMat::operator = (const CovMat& covMat)
{
	this->DeleteAllocatedVar();
	this->Copy(covMat);

	return *this;
}

CovMat operator + (const CovMat& covMat1, const CovMat& covMat2)
{
	return CovMat(*covMat1.matrix + *covMat2.matrix);
}

CovMat operator - (const CovMat& covMat1, const CovMat& covMat2)
{
	return CovMat(*covMat1.matrix - *covMat2.matrix);
}

CovMat operator * (const double mul, const CovMat& covMat)
{
	return CovMat(mul * *covMat.matrix);
}

CovMat operator * (const CovMat& covMat, const double mul)
{
	return CovMat(mul * *covMat.matrix);
}

CovMat operator * (const CovMat& covMat1, const CovMat& covMat2)
{
	return CovMat(*covMat1.matrix * *covMat2.matrix);
}

CovMat operator / (const CovMat& covMat, const double div)
{
	return CovMat(*covMat.matrix / div);
}

CovMat& CovMat::operator += (const CovMat& covMat)
{
	*this->matrix += *covMat.matrix;

	this->DeleteAllocatedVar();
	this->ConstructorInitialize();

	return *this;
}

CovMat& CovMat::operator -= (const CovMat& covMat)
{
	*this->matrix -= *covMat.matrix;

	this->DeleteAllocatedVar();
	this->ConstructorInitialize();

	return *this;
}

CovMat& CovMat::operator *= (const double mul)
{
	*this->matrix *= mul;

	this->DeleteAllocatedVar();
	this->ConstructorInitialize();

	return *this;
}

CovMat& CovMat::operator *= (const CovMat& covMat)
{
	*this->matrix *= *covMat.matrix;

	this->DeleteAllocatedVar();
	this->ConstructorInitialize();

	return *this;
}

CovMat& CovMat::operator /= (const double mul)
{
	*this->matrix /= mul;

	this->DeleteAllocatedVar();
	this->ConstructorInitialize();

	return *this;
}

File: SourceCode/CovMat.py
#!/usr/bin/python
 
import Environment
import numpy
import scipy.linalg

class CovMat :
	# ----------------------------------------------------------------------------------- #
	# ------------------------------- COVMAT CONSTRUCTORS ------------------------------- #
	# ----------------------------------------------------------------------------------- #

	def __init__(self, arg, memorySafeState = Environment.memorySafeState) :
		if (isinstance(arg, int)) :																	#arg is an it
			self.matrixOrder = arg																	#alloc memory only. matrix isn't sym def pos (use randomise() function fot it)
			self.matrix = self.MatrixFromArray(numpy.array((arg, arg)))
			self.FieldsInitialization()
		elif (isinstance(arg, numpy.ndarray)) :														#arg is an ndarray
			self.matrix = self.MatrixFromArray(arg, memorySafeState)										#map an ndarray into a matrix array
			self.matrixOrder = arg.shape[0]
			self.FieldsInitialization()
		elif (isinstance(arg, numpy.matrix)) :														#arg is a matrix
			self.matrix = self.MatrixFromArray(arg, memorySafeState)
			self.matrixOrder = arg.shape[0]
			self.FieldsInitialization()



	@staticmethod
	def MatrixFromArray(numpyArray, memorySafeState = False) :
		return numpy.matrix(numpyArray, Environment.dataType, memorySafeState)



	@staticmethod
	def Zero(matrixOrder) :
		return CovMat(numpy.zeros((matrixOrder, matrixOrder)), False)



	@staticmethod
	def Identity(matrixOrder) :
		return CovMat(numpy.eye(matrixOrder), False)



	@staticmethod
	def Random(matrixOrder) :
		covMat = CovMat(matrixOrder)
		covMat.Randomize()

		return covMat



	def FieldsInitialization(self) :
		self.eigenValues = None
		self.eigenVectors = None
		self.norm = None
		self.determinant = None
		self.inverse = None
		self.sqrtm = None
		self.invsqrtm = None
		self.expm = None
		self.logm = None
		self.powm = None
		self.power = 1



	# ----------------------------------------------------------------------- #
	# ------------------------------- GETTERS ------------------------------- #
	# ----------------------------------------------------------------------- #

	def GetMatrix(self) :
		return self.matrix



	def GetMatrixOrder(self) :
		return self.matrixOrder



	def GetEigenValues(self) :
		if (self.eigenValues is not None) :
			return self.eigenValues
			
		self.ComputeEigen(true)
		return self.eigenValues



	def GetEigenVectors(self) :
		if (self.eigenVectors is not None) :
			return self.eigenVectors
			
		self.ComputeEigen()
		return self.eigenVectors



	def GetEigenVectorsTranspose(self) :
		if (self.eigenVectorsTranspose is not None) :
			return self.eigenVectorsTranspose
			
		self.ComputeEigen()
		return self.eigenVectorsTranspose



	# ------------------------------------------------------------------------------ #
	# ------------------------------- USUAL FUNCTIONS ------------------------------- #
	# ------------------------------------------------------------------------------ #

	def Fill(self, value) :
		self.matrix.fill(value)
		self.FieldsInitialization()



	def Randomize(self) :
		tmp = numpy.random.rand(self.matrixOrder, self.matrixOrder)
		self.matrix = self.MatrixFromArray(numpy.dot(tmp, numpy.transpose(tmp))/100)
		self.FieldsInitialization()



	def Diagonal(self, offset = 0) :
		return self.matrix.diagonal(offset)



	def Column(self, i) :
		return self.matrix[:, i]



	def Row(self, i) :
		return self.matrix[i, :]



	def Maximum(self, axis = None) :
		return self.matrix.max(axis)



	def Minimum(self, axis = None) :
		return self.matrix.min(axis)



	def Mean(self, axis = None) :
		return self.matrix.mean(axis)



	def Variance(self, axis = None) :
		return self.matrix.var(axis)



	def Sum(self, axis = None) :
		return self.matrix.sum(axis)



	def Product(self, axis = None) :
		return self.matrix.prod(axis)



	def Norm(self) :
		if (self.norm is not None) :
			return self.norm

		self.norm = numpy.linalg.norm(self.matrix)
		return self.norm



	def Determinant(self) :
		if (self.determinant is not None) :
			return self.determinant

		self.determinant = numpy.linalg.det(self.matrix)
		return self.determinant



	def Transpose (self) :
		return self



	def Inverse(self) :
		if (self.inverse is not None) :
			return self.inverse

		self.inverse = CovMat(numpy.linalg.inv(self.matrix))
		return self.inverse



	def ComputeEigen(self, eigenValuesOnly = False) :
		if ((self.eigenValues is not None)&(self.eigenVectors is not None)) :
			return
		
		if (eigenValuesOnly) :
			if (self.eigenValues is not None) :
				return

			self.eigenValues = numpy.linalg.eigvalsh(self.matrix)
		else :
			self.eigenValues, self.eigenVectors = numpy.linalg.eigh(self.matrix)
			self.eigenVectors = self.MatrixFromArray(self.eigenVectors)
			self.eigenVectorsTranspose = self.eigenVectors.transpose()



	def Sqrtm (self) :
		if (self.sqrtm is not None) :
			return self.sqrtm

		self.ComputeEigen()
		self.sqrtm = CovMat(self.eigenVectors * self.MatrixFromArray(numpy.diag(numpy.sqrt(self.eigenValues))) * self.eigenVectorsTranspose)
		return self.sqrtm



	def Invsqrtm (self) :
		if (self.invsqrtm is not None) :
			return self.invsqrtm

		self.ComputeEigen()
		self.invsqrtm = CovMat(self.eigenVectors * self.MatrixFromArray(numpy.diag(1.0/numpy.sqrt(self.eigenValues))) * self.eigenVectorsTranspose)
		return self.invsqrtm



	def Expm (self) :
		if (self.expm is not None) :
			return self.expm

		self.ComputeEigen()
		self.expm = CovMat(self.eigenVectors * self.MatrixFromArray(numpy.diag(numpy.exp(self.eigenValues))) * self.eigenVectorsTranspose)
		return self.expm



	def Logm (self) :
		if (self.logm is not None) :
			return self.logm

		self.ComputeEigen()
		self.logm = CovMat(self.eigenVectors * self.MatrixFromArray(numpy.diag(numpy.log(self.eigenValues))) * self.eigenVectorsTranspose)
		return self.logm



	def Powm (self, power) :
		if (power == 1) :
			return self

		if (self.power == power) :
			return self.powm

		self.ComputeEigen()
		self.powm = CovMat(self.eigenVectors * self.MatrixFromArray(numpy.diag(self.eigenValues**power)) * self.eigenVectorsTranspose)
		self.power = power
		return self.powm



	@staticmethod
	def SolveProblem(covMat1, covMat2) :
		return scipy.linalg.eigvalsh(covMat1.matrix, covMat2.matrix)



	# ------------------------------------------------------------------------- #
	# ------------------------------- OPERATORS ------------------------------- #
	# ------------------------------------------------------------------------- #

	def __str__(self) :
		return str(self.matrix)



	def __call__(self, x, y) :
		return self.matrix[x, y]



	def __add__(self, arg) :
		if (isinstance(arg, CovMat)) :
			return CovMat(self.matrix + arg.matrix)
		else :
			return CovMat(self.matrix + arg)



	def __radd__(self, arg) :
		return self.__add__(arg)



	def __iadd__(self, arg) :
		if (isinstance(arg, CovMat)) :
			self.matrix += arg.matrix
		else :
			self.matrix += arg

		self.FieldsInitialization()
		return self



	def __sub__(self, arg) :
		if (isinstance(arg, CovMat)) :
			return CovMat(self.matrix - arg.matrix)
		else :
			return CovMat(self.matrix - arg)



	def __rsub__(self, arg) :
		return CovMat(-1 * self.matrix + arg.matrix)



	def __isub__(self, arg) :
		if (isinstance(arg, CovMat)) :
			self.matrix -= arg.matrix
		else :
			self.matrix -= arg

		self.FieldsInitialization()
		return self



	def __mul__(self, arg) :
		if (isinstance(arg, CovMat)) :
			return CovMat(self.matrix * arg.matrix)
		else :
			return CovMat(self.matrix * arg)



	def __rmul__(self, arg) :
		return self.__mul__(arg)



	def __imul__(self, arg) :
		if (isinstance(arg, CovMat)) :
			self = CovMat(self.matrix * arg.matrix)
		else :
			self = CovMat(self.matrix * arg)

		self.FieldsInitialization()
		return self



	@staticmethod
	def ElementWiseProduct (covMat1, covMat2) :
		return self.MatrixFromArray(numpy.multiply(covMat1, covMat2))



	def __truediv__(self, arg) :
		return CovMat(self.matrix / arg)



	def __rtruediv__(self, arg) :
		return CovMat(arg / self.matrix)



	def __itruediv__(self, arg) :
		self.matrix /= arg
		self.FieldsInitialization()
		return self



	def __pow__(self, arg) :
		return self.Powm(arg)



	def __ipow__(self, arg) :
		self = self.Powm(arg)
		return self
